### DISCUSSION TITLE
GoPro 12 has no GPS, syncing manually is frustrating

### DISCUSSION THREAD

**Discussion by enzimage** [time4tea/gopro-dashboard-overlay#151] [Ideas]
### My suggestion for a feature: 
just like you can use `--video-time-start file-created`
you could have something like `--shift-gpx-timestamp-positive 35` or `--shift-gpx-timestamp-negative 35`
This would interpret the GPX with a 35 second delay or advance and make it much easier to sync video and GPX

### My use case:
I went to a bikepark last week and recorded 21 clips of riding downhill on the GoPro 12.
I also recorded the whole day of activity (about 8 hourse) using a Garmin Fenyx 6.

The problem is that the camera doesn't sync the time correctly. 
It was synced to my phone a few minutes before the ride but it's not in sync with the watch/GPS. 
I also tried syncing it in front of my computer, even using the new timecode function and the time on the camera is not sync'd to the computer or phone.

I could change the timestamp on the file, but then I would have to do the same for all the files. Changing the GPX time will apply for all the files (assuming there is no variation in the GoPro clock, so far this seems to be true).

I used this website to shift the GPX timestamps: https://gotoes.org/strava/Add_Timestamps_To_GPX.php

The problem with using that website is that to sync the clips to the second, there's a lot of back and forth which would be much easier with a simple option in the command line.

I hope this is not too difficult to implement :)

On a side note: This software is fantastic and shows killer skills and determination from the developer(s). It is awesome. I would like to buy you a beer if you have a link for that 🍻

**Comment by enzimage** [comment]
Another issue with the GoPro 12 is that since you're forced to use `--use-gpx-only` you can't use `--load ACCL GRAV CORI` because apparently the former overrides the latter and no data from the GoPro is read.

**Reply by time4tea** [reply]
Thanks for highlighting this use case. I think that the issue is mainly that without GPS data in the GoPro file, we currently don't load anything.

I think I did this mainly because I couldn't see a use case where you'd want it,and also that data used to be aligned by GPS time, which made it tricky to use other data if there was no GPS ... now though, we store the data by frame/video time... so this is do-able.

**Reply by enzimage** [reply]
I didn't see that coming either. I actually bought the Hero 12 and sold my Hero 9 blind to the fact that the new one would have no GPS.

**Reply by time4tea** [reply]
Are you sure it doesn't? I got a hero 11 and was surprised that I didn't get any GPS... but it turns out the default was just to disable it. It was just a setting....
Might be worth double checking?

**Reply by time4tea** [reply]
I just checked the specs.. you are right, and I'm amazed. Such a strange product decision!?!??!

https://gopro.com/en/us/compare

**Comment by time4tea** [comment]
Sorry I should also have replied sooner to say

- this seems like a great idea, thank you!
- thank you for your kind words!

**Reply by enzimage** [reply]
No problem! I'll be happy to provide footage from the GoPro hero 12 if you need it for this or other issues and gpx/fit files to match.

**Comment by enzimage** [comment]
I noticed that on a GoPro Hero 9 video shot at 1080p60 that the ACCL info is ahead of the video by about 40 frames.
You can see here: https://youtu.be/tjQP38HkHhk?si=vcI_TE3Nqul167AD&t=233
(After 2m53s) Z ACCL goes down do zero before the bike leaves the platform and goes up before the bike lands. It should be zero while the bike is falling.

I have to say that the Z ACCL is quite a nice tool for people doing jumps in any sport. It should allow you to measure air time with precision. Say, every time Z ACCL goes below 5 start counting, once it gets above 9.8 stop counting. This could be a "last air time" gauge :)

**Comment by time4tea** [comment]
Thanks for this - ... If you are ok with sharing the file with me I can take a look - If that's ok - please send to gopro-overlay@time4tea.net - You might need to use a file sharing service, I usually suggest WeTransfer.... The file time sync code I thought was now sorted, so this would be great to investigate. Cheers!

**Comment by tarunmaganti** [comment]
I don't know how the software works, and I have Fenix 7 and GoPro 12. I have been using my Fenix extensively for so long that I wanted to record videos of my trips. I didn't know 12 didn't have GPS, and Garmin Virb isn't working properly. I have a time lapse of a cycle trip; how do I use that to sync the video? If it's alright with you, I can also send my video clip and the GPX file of my trip.

**Comment by time4tea** [comment]
Hi - I think that the gist of this thread is that when GoPro videos don't have date-time information in them ( like from a GoPro 12, which has no GPS ) - it can be a bit tricky to sync the video and GPX/FIT information...
You can see that various options we have at the moment here: ... https://github.com/time4tea/gopro-dashboard-overlay/tree/main/docs/bin#create-a-movie-from-gpx-and-video-not-created-with-gopro

**Comment by tarunmaganti** [comment]
Does this video start time and end time work with timelapses too then? The gpx file doesn't have same time difference as the video file.

Regards,
Tarun Maganti
________________________________

**Reply by time4tea** [reply]
Timelapse and timewarp modes are supported in a basic fashion. Did you try it?

**Comment by panjacek** [comment]
I was testing the software today and because i am with DJI using multiple video files on single huge GPX file I had to figure out how to sync them "on fly", then I've realized my camera time drifted by approx 31seconds from garmin gpx time...

1. I've added a hack to pass start_date / end_date to gpx parsing and then simple snippet added to 
gpx_to_timeseries after parsing points
```python    
    if start_date:
        points = [p for p in points if p.dt >= start_date]
    if end_date:
        points = [p for p in points if p.dt <= end_date]
```
It will simply discard all data points outside of start/end dates provided. I think there is still some offset of few hundred ms probably file closure time or something alike. As i see some parsing errors like this:
```
2024-08-25 14:25:39,332][gopro_overlay.log:9]: Request for data at time Timeunit ms=100.0, before start of metadata, returning first item
[2024-08-25 14:25:39,437][gopro_overlay.log:9]: Request for data at time Timeunit ms=200.0, before start of metadata, returning first item
[2024-08-25 14:25:39,549][gopro_overlay.log:9]: Request for data at time Timeunit ms=300.0, before start of metadata, returning first item
Render: 4 [  0%]  [  5.7/s] |                                                                                 | ETA:   0:03:14[2024-08-25 14:25:39,677]
```

2.  I had to manually offset gpx data, but it was just easier to offset end_date by hardcoded 31seconds which looks like my time drift at the time of recording...
```python
                        if args.video_time_end:
                            start_date = fns[args.video_time_end](recording.file) - duration.timedelta() + datetime.timedelta(seconds=-31)
                            end_date = start_date + duration.timedelta()
```
I am still trying to figure out where to add gpx force offset, but app is looking promising.
